--- gtk/ScintillaGTK.cxx.orig	2014-06-03 20:06:09.283128426 +0200
+++ gtk/ScintillaGTK.cxx	2014-06-03 20:07:43.310120943 +0200
@@ -767,7 +767,8 @@
 #endif
 	{
 		// Avoid background drawing flash/missing redraws
-		gtk_widget_set_double_buffered(widtxt, FALSE);
+        /* Keep double buffering on for now, to avoid seg fault bug */
+        /* gtk_widget_set_double_buffered(widtxt, FALSE); */
 	}
 	gtk_widget_set_events(widtxt, GDK_EXPOSURE_MASK);
 	gtk_widget_set_size_request(widtxt, 100, 100);
